driver-core/src/main/com/mongodb/internal/connection/DefaultServer.java

```diff
@@ -220,6 +220,9 @@ public void onResult(final T result, final Throwable t) {
                 invalidate();
                 return (T) e.getResponse();
             } catch (MongoException e) {
+                if (e instanceof MongoSocketException && sessionContext.hasSession()) {
+                    sessionContext.markSessionDirty();
+                }
                 invalidate(e);
```

Comment:
Consider drying up with `invalidate(t, sessionContext);` and sharing with the async.

Reply:
Done